reference/regexp.go

```diff
@@ -53,117 +105,57 @@ var (
+	pathComponent = alphanumeric + zeroOrMore(separator+alphanumeric)
+
+	// remoteName matches the remote-name of a repository. It consists of one
+	// or more forward slash (/) delimited path-components:
+	//
+	//	pathComponent[[/pathComponent] ...] // e.g., "library/ubuntu"
+	remoteName = pathComponent + zeroOrMore(`/`+pathComponent)
+	namePat    = optional(domainAndPort+`/`) + remoteName
+
 	// anchoredNameRegexp is used to parse a name value, capturing the
 	// domain and trailing components.
-	anchoredNameRegexp = regexp.MustCompile(anchoredName)
-
-	referencePat = anchored(capture(namePat),
-		optional(literal(":"), capture(tag)),
-		optional(literal("@"), capture(digestPat)))
-	// ReferenceRegexp is the full supported format of a reference. The regexp
-	// is anchored and has capturing groups for name, tag, and digest
-	// components.
-	ReferenceRegexp = regexp.MustCompile(referencePat)
+	anchoredNameRegexp = regexp.MustCompile(anchored(optional(capture(domainAndPort), `/`), capture(remoteName)))
 
-	identifier = `([a-f0-9]{64})`
-	// IdentifierRegexp is the format for string identifier used as a
-	// content addressable identifier using sha256. These identifiers
-	// are like digests without the algorithm, since sha256 is used.
-	IdentifierRegexp = regexp.MustCompile(identifier)
+	referencePat = anchored(capture(namePat), optional(`:`, capture(tag)), optional(`@`, capture(digestPat)))
 
-	anchoredIdentifier = anchored(identifier)
 	// anchoredIdentifierRegexp is used to check or match an
 	// identifier value, anchored at start and end of string.
-	anchoredIdentifierRegexp = regexp.MustCompile(anchoredIdentifier)
+	anchoredIdentifierRegexp = regexp.MustCompile(anchored(identifier))
 )
 
-// literal compiles s into a literal regular expression, escaping any regexp
-// reserved characters.
-func literal(s string) string {
-	re := regexp.MustCompile(regexp.QuoteMeta(s))
-
-	if _, complete := re.LiteralPrefix(); !complete {
-		panic("must be a literal")
-	}
-
-	return re.String()
-}
-
-// expression defines a full expression, where each regular expression must
-// follow the previous.
-func expression(res ...string) string {
-	var s string
-	for _, re := range res {
-		s += re
-	}
-
-	return s
-}
-
 // optional wraps the expression in a non-capturing group and makes the
 // production optional.
 func optional(res ...string) string {
-	return group(expression(res...)) + `?`
-}
-
-// repeated wraps the regexp in a non-capturing group to get one or more
-// matches.
-func repeated(res ...string) string {
-	return group(expression(res...)) + `+`
+	return `(?:` + strings.Join(res, "") + `)?`
 }
 
-// group wraps the regexp in a non-capturing group.
-func group(res ...string) string {
-	return `(?:` + expression(res...) + `)`
+// zeroOrMore wraps the expression in a non-capturing group that can occur
+// zero or more times.
+func zeroOrMore(res ...string) string {
+	return `(?:` + strings.Join(res, "") + `)*`
 }
```

Comment:
Nowhere is this function called with more than one argument. I would expect that the compiler would be more able to inline and fold down to a single string constant when a `strings.Join` call is not in the way.
```suggestion
func zeroOrMore(res string) string {
	return `(?:` + res + `)*`
}
```

Reply:
I was really on the fence on this one, but in the end decided to keep it consistent with the other options. w.r.t. optimising; these should only run once, so I think the gain would be very minimal.

I'm keeping this one for now, but let me know if you feel strongly about it 👍 


Overall, I really don't like the functions, even considering to change to a generate and put the actual results in a `const`, but I need to give that some thinking. Also want to look if we _really_ need regexes for all of this (some things could possibly be handled with a simpler approach), but tons of improvements that can be made. I started with removing _some_ of the abstractions.

Reply:
They will only run once... in an implicit compiler-generated `func init()`. Given how much of a fuss was made about `func init() { runtime.LockOSThread() }` recently, I'm surprised how casual you are about not optimizing away code _which allocates during init_. Though on the other hand, the upcoming demise of reexec in Moby will move package init out of the hot path, making such optimization concerns mostly moot.

> Also want to look if we _really_ need regexes for all of this (some things could possibly be handled with a simpler approach)

Please, keep all the regexes! This is one of the few places where we build a parser from a formal grammar instead of just having an ad-hoc, hand rolled procedural parser.

Reply:
>  I'm surprised how casual you are about not optimizing away code which allocates during init. 

I probably should've extended a bit on that; yes, we should optimise, but compared to building the regexes itself, I think the `strings.Join()` would be very minor.

I want to do more optimisations following this, but my first goals with these PRs was to bring the package in a "reasonable" shape, to consider moving it separate.